### PR TITLE
Add CI tests for SFT/PEFT

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do ?

Add GitHub actions tests for SFT/PEFT workloads in NeMo 2

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: .github/workflows/cicd-main.yml [hunk 1]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+  L2_NeMo_2_GPT_SFT_TP1PP1_MBS2:
+    needs: [cicd-test-container-setup]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_GPT_SFT_TP1PP1_MBS2') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5266]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 2]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+        --devices 2 \
+        --max-steps 3 \
+        --experiment-dir tests/collections/llm/${{ github.run_id }} \
+        --peft none \
+        --tp_size 1 \
+        --pp_size 1 \
+        --mbs 2
+        
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5276]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 3]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+  L2_NeMo_2_GPT_SFT_TP1PP2_MBS2:
+    needs: [cicd-test-container-setup]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_GPT_SFT_TP1PP2_MBS2') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5297]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 4]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+        --devices 2 \
+        --max-steps 3 \
+        --experiment-dir tests/collections/llm/${{ github.run_id }} \
+        --peft none \
+        --tp_size 1 \
+        --pp_size 2 \
+        --mbs 2
+        
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5307]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 5]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+  L2_NeMo_2_GPT_SFT_TP2PP1_MBS2:
+    needs: [cicd-test-container-setup]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_GPT_SFT_TP2PP1_MBS2') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5328]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 6]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+        --devices 2 \
+        --max-steps 3 \
+        --experiment-dir tests/collections/llm/${{ github.run_id }} \
+        --peft none \
+        --tp_size 2 \
+        --pp_size 1 \
+        --mbs 2
+        
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5338]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 7]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+  L2_NeMo_2_GPT_LoRA_TP1PP1_MBS1:
+    needs: [cicd-test-container-setup]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_GPT_LoRA_TP1PP1_MBS1') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5359]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 8]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+        --devices 2 \
+        --max-steps 3 \
+        --experiment-dir tests/collections/llm/${{ github.run_id }} \
+        --peft lora \
+        --tp_size 1 \
+        --pp_size 1 \
+        --mbs 1
+        
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5369]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 9]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+  L2_NeMo_2_GPT_LoRA_TP1PP1_MBS2:
+    needs: [cicd-test-container-setup]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_GPT_LoRA_TP1PP1_MBS2') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5390]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 10]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+        --devices 2 \
+        --max-steps 3 \
+        --experiment-dir tests/collections/llm/${{ github.run_id }} \
+        --peft lora \
+        --tp_size 1 \
+        --pp_size 1 \
+        --mbs 2
+        
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5400]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 11]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+  L2_NeMo_2_GPT_LoRA_TP1PP2_MBS2:
+    needs: [cicd-test-container-setup]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_GPT_LoRA_TP1PP2_MBS2') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5421]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 12]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+  L2_NeMo_2_GPT_LoRA_TP2PP1_MBS2:
+    needs: [cicd-test-container-setup]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_GPT_LoRA_TP2PP1_MBS2') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5452]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

File: .github/workflows/cicd-main.yml [hunk 13]
```diff
@@ -5223,6 +5223,254 @@ jobs:
...
+        --devices 2 \
+        --max-steps 3 \
+        --experiment-dir tests/collections/llm/${{ github.run_id }} \
+        --peft lora \
+        --tp_size 2 \
+        --pp_size 1 \
+        --mbs 2
+        
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /TestData/nemo2_ckpt/llama_68M
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5462]:
```suggestion
        --restore_path /TestData/nemo2_ckpt/llama_68M \
```

**Comment by cuichenx** [comment]
CI issue should be resolved once #10358 is merged

**Comment by cuichenx** [comment]
New tests passed https://github.com/NVIDIA/NeMo/actions/runs/11223752196/

**Review by akoumpa — DISMISSED** [review]

File: .github/workflows/cicd-main.yml [hunk 14]
```diff
@@ -5183,15 +5183,263 @@ jobs:
...
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_GPT_SFT_TP1PP1_MBS1') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+
+        python tests/collections/llm/gpt_finetuning.py \
+        --restore_path /home/TestData/nemo2_ckpt/llama_68M \
+        --devices 2 \
+        --max_steps 3 \
+        --experiment_dir tests/collections/llm/${{ github.run_id }} \
```
Comment (pablo-garay) [inline .github/workflows/cicd-main.yml:5208]: i see you are using two different paths for the same
--experiment-dir=/tmp/mixtral_pretrain_results \
--experiment_dir tests/collections/llm/${{ github.run_id }} \
Can we unify the idea & use /tmp only ?

**Review by pablo-garay — CHANGES_REQUESTED** [review]

File: .github/workflows/cicd-main.yml (hunk 14 above)
Comment (pablo-garay) [reply .github/workflows/cicd-main.yml:5208]: Discussed @cuichenx will do this in follow up PR

**Review by pablo-garay — APPROVED** [review]

**Comment by pablo-garay** [comment]
CI passed: https://github.com/NVIDIA/NeMo/actions/runs/11296908276/job/31427200519?pr=10632

**Comment by pablo-garay** [comment]
Needed to resolve conflict which had same change. @akoumpa @pablo-garay had already approved & CI passed. Merging...

File: .github/workflows/cicd-main.yml (hunk 14 above)
Comment (cuichenx) [reply .github/workflows/cicd-main.yml:5208]: Here's the follow up PR #10859

**Pull request merged**